rmk/src/split/serial/mod.rs

```diff
@@ -42,44 +42,81 @@ pub(crate) async fn run_serial_peripheral_monitor<
     async fn read(&mut self) -> Result<SplitMessage, SplitDriverError> {
-        let mut buf = [0_u8; SPLIT_MESSAGE_MAX_SIZE];
-        let n_bytes = self
-            .serial
-            .read(&mut buf)
-            .await
-            .map_err(|_e| SplitDriverError::SerialError)?;
-        if n_bytes == 0 {
-            return Err(SplitDriverError::EmptyMessage);
+        const SENTINEL: u8 = 0x00;
+        while self.n_bytes_part < self.buffer.len() {
+            let n_bytes = self
+                .serial
+                .read(&mut self.buffer[self.n_bytes_part..])
+                .await
+                .map_err(|_e| {
+                    self.n_bytes_part = 0;
+                    SplitDriverError::SerialError
+                })?;
+            if n_bytes == 0 {
+                return Err(SplitDriverError::EmptyMessage);
+            }
+
+            self.n_bytes_part = (self.n_bytes_part + n_bytes).min(self.buffer.len());
+            if self.buffer[..self.n_bytes_part].contains(&SENTINEL) {
+                break;
+            }
         }
-        let message: SplitMessage = postcard::from_bytes(&buf).map_err(|e| {
-            error!("Postcard deserialize split message error: {}", e);
-            SplitDriverError::DeserializeError
-        })?;
+
+        let (message, n_bytes_unused) = {
+            let (message, unused_bytes): (SplitMessage, &mut [u8]) =
+                postcard::take_from_bytes_cobs(&mut self.buffer[..self.n_bytes_part]).map_err(
+                    |e| {
+                        error!("Postcard deserialize split message error: {}", e);
+                        self.n_bytes_part = 0;
+                        SplitDriverError::SerializeError
+                    },
+                )?;
+            (message, unused_bytes.len())
+        };
+
+        self.buffer.copy_within(
+            self.n_bytes_part-n_bytes_unused..self.n_bytes_part,
+            0,
+        );
+        self.n_bytes_part = n_bytes_unused;
+
         Ok(message)
     }
 }
 
 impl<S: Read + Write> SplitWriter for SerialSplitDriver<S> {
     async fn write(&mut self, message: &SplitMessage) -> Result<usize, SplitDriverError> {
         let mut buf = [0_u8; SPLIT_MESSAGE_MAX_SIZE];
-        let bytes = postcard::to_slice(message, &mut buf).map_err(|e| {
+        let bytes = postcard::to_slice_cobs(message, &mut buf).map_err(|e| {
             error!("Postcard serialize split message error: {}", e);
             SplitDriverError::SerializeError
         })?;
-        self.serial
-            .write(bytes)
-            .await
-            .map_err(|_e| SplitDriverError::SerialError)
+        let mut remaining_bytes = bytes.len();
+        while remaining_bytes > 0 {
+            let sent_bytes = self
+                .serial
+                .write(&bytes[bytes.len() - remaining_bytes..])
+                .await
+                .map_err(|_e| SplitDriverError::SerialError)?;
+            remaining_bytes -= sent_bytes;
```

Comment:
Is it possible that `sent_bytes` > `remaining_bytes`? If so, `remaining_bytes` could be incorrect. If not, it's safe.

Reply:
Looked at it again, I think it's safe.